File: baml_language/sdks/typescript/bridge_typescript/typescript_src/function_spec.ts
// Host proxy for a live ai.FunctionSpec<Out> value.

import { BamlHandle, getRuntime, newFunctionCall as nativeNewFunctionCall } from './native.js';
import { decodeCallResult, encodeCallArgs } from './proto.js';
import type { BamlPrompt } from './proto.js';
import type { BamlType } from './wire_ty.js';

export interface BamlFunctionSpecCallOptions {
    client?: unknown;
    on_event?: unknown;
}

export interface BamlFunctionSpecBuildRequestOptions {
    client?: unknown;
}

function newFunctionCall(): bigint {
    return BigInt(nativeNewFunctionCall());
}

function suppliedOptions(options: object | undefined): Record<string, unknown> {
    return Object.fromEntries(
        Object.entries(options ?? {}).filter(([, value]) => value !== undefined),
    );
}

/** An opaque, bound LLM recipe owned by the engine that created it. */
export class BamlFunctionSpec<TOut> {
    private readonly handle: BamlHandle;

    constructor(handle: BamlHandle) {
        this.handle = handle;
    }

    /** Internal: construct a FunctionSpec proxy from a tagged heap handle. */
    static _fromHandle<TOut>(
        handle: BamlHandle,
        _classFqn: string,
    ): BamlFunctionSpec<TOut> {
        return new BamlFunctionSpec<TOut>(handle);
    }

    /** Internal: expose the inner handle for inbound encoding. */
    _toHandle(): BamlHandle {
        return this.handle;
    }

    name(): string {
        return this._callSync('ai.FunctionSpec.name') as string;
    }

    async nameAsync(): Promise<string> {
        return await this._callAsync('ai.FunctionSpec.name') as string;
    }

    arguments(): Record<string, unknown> {
        return this._callSync('ai.FunctionSpec.arguments') as Record<string, unknown>;
    }

    async argumentsAsync(): Promise<Record<string, unknown>> {
        return await this._callAsync('ai.FunctionSpec.arguments') as Record<string, unknown>;
    }

    outputType(): BamlType {
        return this._callSync('ai.FunctionSpec.output_type') as BamlType;
    }

    async outputTypeAsync(): Promise<BamlType> {
        return await this._callAsync('ai.FunctionSpec.output_type') as BamlType;
    }

    prompt(): BamlPrompt {
        return this._callSync('ai.FunctionSpec.prompt') as BamlPrompt;
    }

    async promptAsync(): Promise<BamlPrompt> {
        return await this._callAsync('ai.FunctionSpec.prompt') as BamlPrompt;
    }

    tools(): unknown {
        return this._callSync('ai.FunctionSpec.tools');
    }

    async toolsAsync(): Promise<unknown> {
        return await this._callAsync('ai.FunctionSpec.tools');
    }

    clientId(): string {
        return this._callSync('ai.FunctionSpec.client_id') as string;
    }

    async clientIdAsync(): Promise<string> {
        return await this._callAsync('ai.FunctionSpec.client_id') as string;
    }

    buildRequest(options?: BamlFunctionSpecBuildRequestOptions): unknown {
        return this._callSync('ai.FunctionSpec.build_request', suppliedOptions(options));
    }

    async buildRequestAsync(options?: BamlFunctionSpecBuildRequestOptions): Promise<unknown> {
        return await this._callAsync(
            'ai.FunctionSpec.build_request',
            suppliedOptions(options),
        );
    }

    parse(json: string): TOut {
        return this._callSync('ai.FunctionSpec.parse', { json }) as TOut;
    }

    async parseAsync(json: string): Promise<TOut> {
        return await this._callAsync('ai.FunctionSpec.parse', { json }) as TOut;
    }

    call(options?: BamlFunctionSpecCallOptions): TOut {
        return this._callSync('ai.FunctionSpec.call', suppliedOptions(options)) as TOut;
    }

    async callAsync(options?: BamlFunctionSpecCallOptions): Promise<TOut> {
        return await this._callAsync(
            'ai.FunctionSpec.call',
            suppliedOptions(options),
        ) as TOut;
    }

    private _callSync(fqn: string, kwargs: Record<string, unknown> = {}): unknown {
        const argsProto = encodeCallArgs(
            { self: this, ...kwargs },
            { syncMode: true, callId: newFunctionCall(), functionName: fqn },
        );
        return decodeCallResult(getRuntime().callFunctionSync(argsProto, null, null));
    }

    private async _callAsync(
        fqn: string,
        kwargs: Record<string, unknown> = {},
    ): Promise<unknown> {
        const argsProto = encodeCallArgs(
            { self: this, ...kwargs },
            { callId: newFunctionCall(), functionName: fqn },
        );
        return decodeCallResult(await getRuntime().callFunction(argsProto, null, null));
    }

    toString(): string {
        return '<BamlFunctionSpec>';
    }
}

File: baml_language/sdks/python/src/baml_bridge/_function_spec.py
"""Host proxy for a live ``ai.FunctionSpec<Out>`` value."""

from __future__ import annotations

import asyncio
from typing import Any, Generic, TypeVar

from .typemap import get_type_map, runtime_bound
from .baml_py import BamlPyHandle

TOut = TypeVar("TOut")


class BamlFunctionSpec(Generic[TOut]):
    """Opaque bound LLM recipe owned by the engine that created it."""

    __slots__ = ("_type_map", "_handle",)

    def __init__(self, handle: BamlPyHandle) -> None:
        self._type_map = get_type_map()
        self._handle = handle

    @classmethod
    def _from_pyhandle(cls, handle: BamlPyHandle) -> "BamlFunctionSpec[Any]":
        return cls(handle)

    def _to_pyhandle(self) -> BamlPyHandle:
        return self._handle

    def name(self) -> str:
        return self._call_sync("ai.FunctionSpec.name")

    async def name_async(self) -> str:
        return await self._call_async("ai.FunctionSpec.name")

    def arguments(self) -> dict[str, Any]:
        return self._call_sync("ai.FunctionSpec.arguments")

    async def arguments_async(self) -> dict[str, Any]:
        return await self._call_async("ai.FunctionSpec.arguments")

    def output_type(self) -> Any:
        return self._call_sync("ai.FunctionSpec.output_type")

    async def output_type_async(self) -> Any:
        return await self._call_async("ai.FunctionSpec.output_type")

    def prompt(self) -> Any:
        return self._call_sync("ai.FunctionSpec.prompt")

    async def prompt_async(self) -> Any:
        return await self._call_async("ai.FunctionSpec.prompt")

    def tools(self) -> Any:
        return self._call_sync("ai.FunctionSpec.tools")

    async def tools_async(self) -> Any:
        return await self._call_async("ai.FunctionSpec.tools")

    def client_id(self) -> str:
        return self._call_sync("ai.FunctionSpec.client_id")

    async def client_id_async(self) -> str:
        return await self._call_async("ai.FunctionSpec.client_id")

    def build_request(self, **kwargs: Any) -> Any:
        return self._call_sync("ai.FunctionSpec.build_request", kwargs)

    async def build_request_async(self, **kwargs: Any) -> Any:
        return await self._call_async("ai.FunctionSpec.build_request", kwargs)

    def parse(self, json: str) -> TOut:
        return self._call_sync("ai.FunctionSpec.parse", {"json": json})

    async def parse_async(self, json: str) -> TOut:
        return await self._call_async("ai.FunctionSpec.parse", {"json": json})

    def call(self, **kwargs: Any) -> TOut:
        return self._call_sync("ai.FunctionSpec.call", kwargs)

    async def call_async(self, **kwargs: Any) -> TOut:
        return await self._call_async("ai.FunctionSpec.call", kwargs)

    @runtime_bound
    def _call_sync(self, fqn: str, kwargs: dict[str, Any] | None = None) -> Any:
        from . import get_runtime
        from .baml_py import new_function_call
        from .proto import decode_call_result, encode_call_args

        values = {"self": self}
        values.update(kwargs or {})
        encoded = encode_call_args(
            values,
            new_function_call(),
            function_name=fqn,
        )
        return decode_call_result(get_runtime().call_function_sync(encoded, None, None))

    @runtime_bound
    async def _call_async(self, fqn: str, kwargs: dict[str, Any] | None = None) -> Any:
        from . import _decode_call_result_async, cancel_function_call, get_runtime
        from .baml_py import new_function_call
        from .proto import encode_call_args

        values = {"self": self}
        values.update(kwargs or {})
        call_id = new_function_call()
        encoded = encode_call_args(values, call_id, function_name=fqn)
        try:
            result = await get_runtime().call_function(encoded, None, None)
        except asyncio.CancelledError:
            try:
                cancel_function_call(call_id)
            except Exception:
                pass
            raise
        return _decode_call_result_async(result)

    @classmethod
    def __get_pydantic_core_schema__(cls, _source_type: Any, _handler: Any) -> Any:
        from pydantic_core import core_schema  # type: ignore[import-untyped]

        return core_schema.is_instance_schema(cls)

    def __repr__(self) -> str:
        return "<BamlFunctionSpec>"
